clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -2103,6 +2104,144 @@ static void handleCHERIMethodSuffix(Sema &S, Decl *D, const ParsedAttr &Attr) {
   D->addAttr(::new (S.Context) CHERIMethodSuffixAttr(S.Context, Attr, Str));
 }
 
+static void handleCHERIMMIODevice(Sema &S, Decl *D, const ParsedAttr &Attr,
+                                  Sema::DeclAttributeLocation DAL) {
+  StringRef DeviceName;
+  SourceLocation DeviceNameLiteralLoc;
+  if (!S.checkStringLiteralArgumentAttr(Attr, 0, DeviceName,
+                                        &DeviceNameLiteralLoc))
+    return;
+  StringRef Permissions;
+  SourceLocation PermissionsLiteralLoc;
+
+  if (Attr.getNumArgs() > 1) {
+    S.checkStringLiteralArgumentAttr(Attr, 1, Permissions,
+                                     &PermissionsLiteralLoc);
+  }
+
+  if (Permissions.empty()) {
+    return D->addAttr(::new (S.Context) CHERIMMIODeviceAttr(
+        S.Context, Attr, DeviceName,
+        llvm::cheri::GlobalCapabilityImportAttr::defaultPermissions()));
+  }
+
+  auto ScratchPermissions = Permissions.str();
+  const auto *ValidPermissionSymbols =
+      llvm::cheri::GlobalCapabilityImportAttr::getValidPermissionSymbols();
+
+  for (char Symbol : *ValidPermissionSymbols) {
+
+    int SymbolOccurrences = 0;
+    auto new_end = std::remove_if(ScratchPermissions.begin(),
```

Comment:
This seems like a somewhat round-about way of implementing this check. I would instead formulate it as a fold-like loop, where the accumulator is a bit-mask of already seen permissions.

Reply:
(I don't know if C++ `<functional>` has good support for this. Personally I'd just write a for loop, but I'm old-school sometimes.

Reply:
https://github.com/CHERIoT-Platform/llvm-project/pull/147/commits/58fd6c5e4c2bab83ad8dc05fd32fc0ac3f0a1872#diff-9bdc401a1ee7793838a466dbec73b4cd8ce3a57e222b19e06fc9d429d2c561c3R224-R255 rewrites it in a (hopefully) clearer way.